### PR TITLE
FeatureCounts: added profile

### PR DESCRIPTION
**Pull request by yhoogstrate**
This tool was not deployed because it was only compatible with galaxy 16.04 and higher. As we've now reached that point I think it's time to deploy it.

- [x] Update to latest tool standards

### CODE REVIEW

**Comment by bgruening** [comment]
@yhoogstrate if we use the profile feature we should migrate the tool to all of the new features, e.g. remove the stdio tags.

File: tools/featurecounts/featurecounts.xml [hunk 1]
```diff
@@ -1,13 +1,9 @@
...
     <description>Measure gene expression in RNA-Seq experiments from SAM or BAM files.</description>
     <requirements>
         <requirement type="package" version="1.4.6p5">subread</requirement>
     </requirements>
-    <stdio>
-        <!-- Anything other than zero is an error -->
-        <exit_code range="1:" />
-        <exit_code range=":-1" />
-    </stdio>
+
```
Comment (bgruening) [inline tools/featurecounts/featurecounts.xml]: Just as information, I don't think it is deprecated it has a more sane default setting so we do not need to use it so often anymore.

**Comment by yhoogstrate** [comment]
Are there other changes since Galaxy 16.04? I can't find other deprecations than those described in https://github.com/galaxyproject/galaxy/pull/1688 so it should now be fine now?

**Comment by shiltemann** [comment]
@yhoogstrate you have two conditionals with booleans, should those be changed to selects? http://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html?highlight=conditionals#booleans

**Comment by yhoogstrate** [comment]
Is this still the best practice after https://github.com/galaxyproject/galaxy/pull/1714? I can imagine that in combination with profile="16.04" or higher this constraint is not necessary anymore?

**Comment by bgruening** [comment]
Afaik this was more a design decision to have a coherent interface for all tools and not mixing booleans with selects.

**Comment by shiltemann** [comment]
IE doesn't show the extra parameters if it is a boolean so think that was why this is in the best practices

File: tools/featurecounts/featurecounts.xml [hunk 2]
```diff
@@ -123,6 +119,76 @@
...
+                       help="If specified, fragments (or templates) will be counted instead of reads.">
+                    <option value="" selected="true">Disabled; all reads/mates will be counted individually</option>
+                    <option value=" -p">Enabled; fragments (or templates) will be counted instead of reads</option>
+                </param>
+
+                <when value=" -p">
+                    <conditional name="check_distance_enabled">
+                        <param name="check_distance"
+                            type="boolean"
+                            truevalue=" -P"
```
Comment (bgruening) [inline tools/featurecounts/featurecounts.xml]: This should be a select and the space in the truevalue looks wired. Is this really needed?

**Comment by pavanvidem** [comment]
@yhoogstrate I wonder why all counts are written to a single tabular file, if multiple input samples (or collections) are used? How this output can be used further? One problem is that the current DESeq2 wrapper takes individual count table for each sample. Hence I need to remove the header and split the counts manually from the current output format.

**Comment by yhoogstrate** [comment]
Dear @pavanvidem,

This is more a legacy thing, and was part of a never finished discussion between @mvdbeek @bgruening and me. By the time of writing the initial wrapper there was no properly working collection system and I was using the featureCounts matrices for an edgeR wrapper which does use such tabular data files. The main rational behind it was that the same matrix could be used for testing multiple hypotheses because the counts are static. This implementation saved a lot of clicking for certain complex experimental designs with many samples. With the same matrix you could do e.g. paired analysis, swap design matrix for a totally different test, but also estimate dispersion over samples that are not taken into account for actual testing.

Anyway, things have changed and improved really well, especially with respect to the collections. If we're now sure we're not heading in the direction of using large matrices I'm fine with adding `multiple="false"` to the `alignments` variable and to clean-up the loop and test-file(s).

I would rather see the DESeq2 wrapper to be compatible with one header line if the first lines' second column is non-numerical, but if you would prefer to remove it here we can do that as well. If you use non headered columns outside of galaxy by downloading, it can be confusing to only have galaxy's file name to track the sample name back.

Thanks for looking into this, it is really appreciated :+1: .

**Comment by pavanvidem** [comment]
Hi @yhoogstrate, 
Thanks for your explanation. I understand the problem now. Currently we are using dataset collections to deal with large experiments with several conditions instead of matrices. Hence, any option that gives collection of count tables as output would be nice. In that case, I'll try to adapt the DESeq2 wrapper to deal with header. Thanks!

**Comment by bgruening** [comment]
> I would rather see the DESeq2 wrapper to be compatible with one header line if the first lines' second column is non-numerical, but if you would prefer to remove it here we can do that as well. If you use non headered columns outside of galaxy by downloading, it can be confusing to only have galaxy's file name to track the sample name back, that's why I like to have a header row.

I still think headers in tabular files are confusing and hard to use for people using Galaxy. If the only concern is for people that downloading datasets, we should fix this on a framework level and write the metadata to the output file (optionally). I think this can be done.

**Comment by yhoogstrate** [comment]
I think the wrapper now produces tables that are compatible with the DESeq2 wrapper and produces 1 output file per sample. Column names are set as meta data instead of being part of the file:

screenshot_20160906_095333

**Comment by bgruening** [comment]
@pavanvidem can you please look at this?

**Comment by bgruening** [comment]
@yhoogstrate this looks great thanks a lot!

**Comment by lparsons** [comment]
@yhoogstrate This looks really nice, I've been wanting a version of featureCounts that works with DESeq2. Is there still a way to produce a single output file with all samples (for users of other downstream tools, etc.)? Or does anyone know of a simple way to do this for a collection of results? Thanks.

**Comment by yhoogstrate** [comment]
@lparsons using the tools `cut` and `paste`? I am now realizing that those tools probably don't merge the column names within the meta-data properly... 

ping @pavanvidem would you like to comment on all this? I think we're almost there and I've had several requests to get featureCounts in galaxy

**Comment by bgruening** [comment]
@pavanvidem seems to be busy at the moment. Feel free to merge this! Great work!

**Comment by mvdbeek** [comment]
Very happy about this, thanks a lot @yhoogstrate !

**Comment by lparsons** [comment]
@yhoogstrate Hmmm... I'm a bit surprised that merging files in a collection just doesn't seem possible at the moment in Galaxy. If this tool isn't going to support a combined tabular output, then perhaps I'll need to work on a version of join that will reduce a collection. Thoughts anyone?

**Comment by bgruening** [comment]
@lparsons happy to accept any patches for the https://github.com/bgruening/galaxytools/tree/master/tools/text_processing/text_processing suite.

**Comment by yhoogstrate** [comment]
If bjoerns text Processing tools are way better than galaxy's internal ones we might consider committing them to Galaxy Straight ahead or removing the ones within Galaxy, to avoid confusion

**Comment by bgruening** [comment]
I would't say they are better. They are to 90% pure Unix based and more complete/complex in many cases. I think the consensus was that Galaxy will stick to the default ones for ever, but the unix-ones should be used in parallel. @blankenberg do I remember correctly?

**Comment by lparsons** [comment]
One issue I had was that, for example, @bgruening cat would reduce a list of files, while the internal cat would run in parallel. Depending on the application, both are useful, but it's hard to know or see what the difference is. A better tool would give an option for how to handle lists/collections.

I'm open to making some patches, just trying to maximize time. If this version of featureCounts output a table of results from a collection as an option, that would be better (simpler for the user) and would be easier for me (no need to modify the multi-join tool). So, I was hoping things would go that way.

**Comment by yhoogstrate** [comment]
@bgruening it was not automatically being pushed to the MTS. Is this because it needs a rebase to a more up2date `.travis.yml` and should I push manually, or do I need to be worried?

**Comment by blankenberg** [comment]
In general, I don't think we are looking to add any new tools to the framework at this point. 

Now the question of Unix vs. Other based tools for the text processing gets complicated. Consider the cut tool as an example, Galaxy's cut tool allows reordering and multiple occurrences of columns, whereas GNU cut does not. I'd argue that Galaxy's version of cut is significantly more useful. In cases where functionality is equivalent, it does make sense to use standard *nix tools under the hood.

If you are attempting to join several tabular files into a single tabular file, there are a couple approaches available. @bgruening has one, and I also made one (https://toolshed.g2.bx.psu.edu/view/iuc/collection_column_join/2f120a5c49b1  / https://github.com/galaxyproject/tools-iuc/blob/master/tools/collection_column_join/collection_column_join.xml).

**Comment by pavanvidem** [comment]
I wanted test this. Sorry, but I was bit busy. I saw the screenshot of  the output and that's exactly what we want for DESeq2. Thanks a lot!

**Comment by lparsons** [comment]
@blankenberg The collection column join tool is *exactly* what I was looking for, thanks! 😄

**Pull request merged**